### PR TITLE
Test boot aggregate

### PR DESCRIPTION
**Pull request by Krishan-Saraswat**
localhost:# avocado run --max-parallel-tasks=1 grub-extend-pcr.py
JOB ID     : 1dfd2c3459f708c42271d21a32e6251d5706d9d4
JOB LOG    : /root/avocado-fvt-wrapper/results/job-2024-01-16T00.36-1dfd2c3/job.log
 (1/4) grub-extend-pcr.py:GrubExtendPCR.test_tsspcrread_8: STARTED
 (1/4) grub-extend-pcr.py:GrubExtendPCR.test_tsspcrread_8: PASS (0.04 s)
 (2/4) grub-extend-pcr.py:GrubExtendPCR.test_tsspcrread_9: STARTED
 (2/4) grub-extend-pcr.py:GrubExtendPCR.test_tsspcrread_9: PASS (0.04 s)
 (3/4) grub-extend-pcr.py:GrubExtendPCR.test_tsseventextend: STARTED
 (3/4) grub-extend-pcr.py:GrubExtendPCR.test_tsseventextend: PASS (0.04 s)
 (4/4) grub-extend-pcr.py:GrubExtendPCR.test_boot_aggregate: STARTED
 (4/4) grub-extend-pcr.py:GrubExtendPCR.test_boot_aggregate: PASS (0.04 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado-fvt-wrapper/results/job-2024-01-16T00.36-1dfd2c3/results.html
JOB TIME   : 30.53 s

### CODE REVIEW

File: workload/ebizzy.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: 2022 IBM
```
Comment (Naresh-ibm) [inline workload/ebizzy.py:14]: Please change this to 2024

File: workload/ebizzy.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+# Author: Disha Goel <disgoel@linux.vnet.ibm.com>
```
Comment (Naresh-ibm) [inline workload/ebizzy.py:15]: Author name also

File: workload/ebizzy.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        for package in deps:
+            if not smm.check_installed(package) and not smm.install(package):
+                self.cancel('%s is needed for the test to be run' % package)
+
+        # Creating a temprory file
+        self.temp_file = tempfile.NamedTemporaryFile().name
+
+    def test_workload(self):
+        # Get ebizzy workload and build
+        url = 'https://sourceforge.net/projects/ebizzy/files/ebizzy/0.3/ebizzy-0.3.tar.gz'
```
Comment (Naresh-ibm) [inline workload/ebizzy.py:64]: is this possible to get it from yaml?

File: workload/ebizzy.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                                         shell=True,
+                                         stderr=subprocess.STDOUT)
+        if not output:
+            self.fail("ebizzy workload not captured in perf top")
+
+    def run_workload(self):
+        process.run("./ebizzy -S1 -s1024 -t10", shell=True)
+
+    def capture_top_output(self):
+        process.getoutput("perf top -a > %s " % self.temp_file, timeout=10)
```
Comment (Naresh-ibm) [inline workload/ebizzy.py:93]: use "process.run"

**Review by Naresh-ibm — CHANGES_REQUESTED** [review]

File: workload/ebizzy.py (hunk 4 above)
Comment (Krishan-Saraswat) [reply workload/ebizzy.py]: updated the file with latest changes

**Comment by Naresh-ibm** [comment]
@Krishan-Saraswat the files you have pushed are different.
can you check and confirm

**Comment by Krishan-Saraswat** [comment]
> @Krishan-Saraswat the files you have pushed are different. can you check and confirm

@Naresh-ibm I have fixed the commits that happened because I didn't move the branch to master and two different commits got combined.

File: security/grub-extend-pcr.py [hunk 5]
```diff
@@ -91,3 +91,21 @@ def test_tsseventextend(self):
...
+    def test_boot_aggregate(self):
+        '''
+        Output validation for boot aggregate from two different files
+        ascii_runtime_measurements and binary_bios_measurements
+        '''
+        if not os.path.exists("/sys/kernel/security/ima/ascii_runtime_measurements"):
+            self.fail("files doesn't exist")
+        if not os.path.exists("/sys/kernel/security/tpm0/binary_bios_measurements"):
+            self.fail("file doesn't exist")
+        cmd1 = "head -n1 /sys/kernel/security/ima/ascii_runtime_measurements"
```
Comment (nasastry) [inline security/grub-extend-pcr.py:104]: Please use the variable name for the filename(s). You used the filename more than once.

File: security/grub-extend-pcr.py [hunk 6]
```diff
@@ -91,3 +91,21 @@ def test_tsseventextend(self):
...
+        cmd2 = "tsseventextend -if /sys/kernel/security/tpm0/binary_bios_measurements -sim -pcrmax 9"
```
Comment (nasastry) [inline security/grub-extend-pcr.py:105]: Please use the variable name for the filename(s). You used the filename more than once.

File: security/grub-extend-pcr.py [hunk 7]
```diff
@@ -91,3 +91,21 @@ def test_tsseventextend(self):
...
         if not (pcr8_flag and pcr9_flag):
             self.fail("PCR 8 and/or PCR 9 not having correct values.")
+
+    def test_boot_aggregate(self):
+        '''
+        Output validation for boot aggregate from two different files
+        ascii_runtime_measurements and binary_bios_measurements
+        '''
+        if not os.path.exists("/sys/kernel/security/ima/ascii_runtime_measurements"):
+            self.fail("files doesn't exist")
```
Comment (nasastry) [inline security/grub-extend-pcr.py:101]: Please mention what file is not there.

File: security/grub-extend-pcr.py [hunk 8]
```diff
@@ -91,3 +91,21 @@ def test_tsseventextend(self):
...
+        if not os.path.exists("/sys/kernel/security/tpm0/binary_bios_measurements"):
+            self.fail("file doesn't exist")
```
Comment (nasastry) [inline security/grub-extend-pcr.py:103]: Please mention what file is not there.

File: security/grub-extend-pcr.py [hunk 9]
```diff
@@ -91,3 +91,21 @@ def test_tsseventextend(self):
...
+        cmd1 = "head -n1 /sys/kernel/security/ima/ascii_runtime_measurements"
+        cmd2 = "tsseventextend -if /sys/kernel/security/tpm0/binary_bios_measurements -sim -pcrmax 9"
+        arm_output = process.system_output(cmd1, ignore_status=True).decode()
+        arm_value = arm_output.split(" ")[-2].split(":")[1]
+        tssevent_output = process.system_output(cmd2, ignore_status=True).decode().splitlines()[-1]
+        tssevent_value = tssevent_output.split(":")[1].strip().replace(" ", "")
+        if arm_value != tssevent_value:
+            self.fail("Boot aggregate output doesn't match")
```
Comment (nasastry) [inline security/grub-extend-pcr.py:111]: Improve this message - 'Boot aggregate output not matched from ascii and binary measurements'.

**Review by nasastry — CHANGES_REQUESTED** [review]

File: security/grub-extend-pcr.py (hunk 5 above)
Comment (Krishan-Saraswat) [reply security/grub-extend-pcr.py]: resolved

File: security/grub-extend-pcr.py (hunk 6 above)
Comment (Krishan-Saraswat) [reply security/grub-extend-pcr.py]: resolved

File: security/grub-extend-pcr.py (hunk 7 above)
Comment (Krishan-Saraswat) [reply security/grub-extend-pcr.py]: resolved

File: security/grub-extend-pcr.py (hunk 8 above)
Comment (Krishan-Saraswat) [reply security/grub-extend-pcr.py]: resolved

File: security/grub-extend-pcr.py (hunk 9 above)
Comment (Krishan-Saraswat) [reply security/grub-extend-pcr.py]: resolved.

**Review by PraveenPenguin — APPROVED** [review]

**Review by nasastry — APPROVED** [review]

**Review by Naresh-ibm — APPROVED** [review]
@Krishan-Saraswat Thanks for PR
@PraveenPenguin @nasastry  Thanks for review
LGTM

**Comment by Naresh-ibm** [comment]
@PraveenPenguin @nasastry Thanks for review.

**Pull request merged**